Sources/Basic/Path.swift

```diff
@@ -0,0 +1,473 @@
+    /// Last path component (including the suffix, if any).  it is never empty.
+    public var basename: String {
+        return _impl.basename
+    }
+    
+    /// Suffix (including leading `.` character) if any.  Note that a basename
+    /// that starts with a `.` character is not considered a suffix.
+    public var suffix: String? {
+        return _impl.suffix
+    }
+    
+    /// Normalized string representation (the normalization rules are described
+    /// in the documentation of the initializer).  This string is never empty.
+    public var asString: String {
+        return _impl.string
+    }
+}
+
+
+// Make absolute paths Hashable.
+extension AbsolutePath : Hashable {
+    public var hashValue: Int {
+        return self.asString.hashValue
+    }
+}
+
+// Make absolute paths Equatable.
+extension AbsolutePath : Equatable { }
+public func ==(lhs: AbsolutePath, rhs: AbsolutePath) -> Bool {
+    return lhs.asString == rhs.asString
+}
+
+
+// Make relative paths Hashable.
+extension RelativePath : Hashable {
+    public var hashValue: Int {
+        return self.asString.hashValue
+    }
+}
+
+// Make relative paths Equatable.
+extension RelativePath : Equatable { }
+public func ==(lhs: RelativePath, rhs: RelativePath) -> Bool {
+    return lhs.asString == rhs.asString
+}
+
+
+/// Private implementation shared between AbsolutePath and RelativePath.  It is
+/// a little unfortunate that there needs to be duplication at all between the
+/// AbsolutePath and RelativePath struct, but PathImpl helps mitigate it.  From
+/// a type safety perspective, absolute paths and relative paths are genuinely
+/// different.
+private struct PathImpl {
+    /// Normalized string of the (absolute or relative) path.  Never empty.
+    private let string: String
+    
+    /// Private function that returns the directory part of the stored path
+    /// string (relying on the fact that it has been normalized).  Returns nil
+    /// if there is no directory part (which is the case iff there is no path
+    /// separator).
+    private var dirname: String? {
+        let chars = string.characters
+        guard let idx = chars.rindex(of: PathSeparatorCharacter) else {
+            return nil
+        }
+        return idx == chars.startIndex ? String(chars[idx])
+            : String(chars.prefix(upTo: idx))
+    }
+    
+    private var basename: String {
+        // FIXME: This needs to be rewritten.
+        guard !string.isEmpty else { return "." }
+        let parts = string.characters.split(separator: PathSeparatorCharacter)
+        guard !parts.isEmpty else { return "/" }
+        return String(parts.last!)
+    }
+    
+    private var suffix: String? {
+        // FIXME: This needs to be rewritten.
+        let chars = string.characters
+        guard chars.contains(".") else { return nil }
+        let parts = chars.split(separator: ".")
+        if let last = parts.last where parts.count > 1 { return String(last) }
+        return nil
+    }
```

Comment:
Right.  I forgot to address the `FIXME: This needs to be rewritten` (this code came from the old Path implementation).  I'll fix it.